src/Storages/ExternalStream/Kafka/KafkaSink.cpp

```diff
@@ -20,37 +21,39 @@ extern const int CANNOT_WRITE_TO_KAFKA;
 extern const int MISSING_ACKNOWLEDGEMENT;
 extern const int INVALID_CONFIG_PARAMETER;
 extern const int TYPE_MISMATCH;
+extern const int INVALID_SETTING_VALUE;
 }
 
-namespace KafkaStream
+namespace
 {
-void ChunkSharder::useRandomSharding()
+ExpressionActionsPtr buildExpression(ContextPtr context, const Block & header, const ASTPtr & expr_ast)
 {
-    random_sharding = true;
-    std::random_device r;
-    rand = std::minstd_rand(r());
+    assert(expr_ast);
+
+    ASTPtr query = expr_ast;
+    auto syntax_result = TreeRewriter(context).analyze(query, header.getNamesAndTypesList());
+    return ExpressionAnalyzer(query, syntax_result, context).getActions(false);
+}
 }
 
+namespace KafkaStream
+{
 ChunkSharder::ChunkSharder(ContextPtr context, const Block & header, const ASTPtr & sharding_expr_ast)
 {
-    assert(sharding_expr_ast);
-
-    ASTPtr query = sharding_expr_ast;
-    auto syntax_result = TreeRewriter(context).analyze(query, header.getNamesAndTypesList());
-    sharding_expr = ExpressionAnalyzer(query, syntax_result, context).getActions(false);
+    sharding_expr = buildExpression(context, header, sharding_expr_ast);
```

Comment:
Since Context and Block header are used only once in ctor. We may simplify ChunkSharder to have less dependency and i think it is time to refactor it out in different source files (next PR)


```
ChunkSharder::ChunkSharder(ExpressionActionsPtr sharding_expr_, const ASTPtr & sharding_expr_ast)
    : sharding_expr(std::move(sharding_expr_))
{
  ...
}
```

On call side, we do 

```
ChunkSharder sharder(buildExpression(context, header, sharding_expr_ast), sharding_expr_ast);
```